tremor-script/src/event_payload.rs

```diff
@@ -0,0 +1,278 @@
+    /// of the struct.
+    #[must_use]
+    pub fn new<F>(raw: Vec<u8>, f: F) -> Self
+    where
+        F: for<'head> FnOnce(&'head mut [u8]) -> ValueAndMeta<'head>,
+    {
+        let mut raw = Pin::new(raw);
+        let structured = f(raw.as_mut().get_mut());
+        // This is where the magic happens
+        // ALLOW: this is sound since we implement a self referential struct
+        let structured: ValueAndMeta<'static> = unsafe { mem::transmute(structured) };
+        let raw = vec![Arc::new(raw)];
+        Self { raw, structured }
+    }
+
+    /// Creates a new Payload with a given byte vector and
+    /// a function to turn it into a value and metadata set.
+    ///
+    /// The return can reference the the data it gets passed
+    /// in the function.
+    ///
+    /// Internally the lifetime will be bound to the raw part
+    /// of the struct.
+    ///
+    /// # Errors
+    /// errors if the conversion function fails
+    pub fn try_new<E, F>(raw: Vec<u8>, f: F) -> std::result::Result<Self, E>
+    where
+        F: for<'head> FnOnce(&'head mut [u8]) -> std::result::Result<ValueAndMeta<'head>, E>,
+    {
+        let mut raw = Pin::new(raw);
+        let structured = f(raw.as_mut().get_mut())?;
+        // This is where the magic happens
+        // ALLOW: this is sound since we implement a self referential struct
+        let structured: ValueAndMeta<'static> = unsafe { mem::transmute(structured) };
+        let raw = vec![Arc::new(raw)];
+        Ok(Self { raw, structured })
+    }
+
+    /// Named after the original rental struct for easy rewriting.
+    ///
+    /// Borrows the borrowed (liftimed) part of the self referential struct
+    /// and calls the provided function with a reference to it
+    pub fn rent<F, R>(&self, f: F) -> R
+    where
+        F: for<'iref, 'head> FnOnce(&'iref ValueAndMeta<'head>) -> R,
+        R: ,
+    {
+        f(&self.structured)
+    }
+
+    /// Named after the original rental struct for easy rewriting.
+    ///
+    /// Borrows the borrowed (liftimed) part of the self referential struct
+    /// mutably and calls the provided function with a mutatable reference to it
+    pub fn rent_mut<F, R>(&mut self, f: F) -> R
+    where
+        F: for<'iref, 'head> FnOnce(&'iref mut ValueAndMeta<'head>) -> R,
+        R: ,
+    {
+        f(&mut self.structured)
+    }
+
+    /// Borrow the parts (event and metadata) from a rental.
+    #[must_use]
+    pub fn parts<'value, 'borrow>(&'borrow self) -> (&'borrow Value<'value>, &'borrow Value<'value>)
+    where
+        'borrow: 'value,
+    {
+        let ValueAndMeta { ref v, ref m } = self.structured;
+        (v, m)
+    }
+
+    /// Consumes one payload into another
+    ///
+    /// # Errors
+    /// if `join_f` errors
+    pub fn consume<E, F>(&mut self, mut other: Self, join_f: F) -> Result<(), E>
+    where
+        E: std::error::Error,
+        F: Fn(&mut ValueAndMeta<'static>, ValueAndMeta<'static>) -> Result<(), E>,
```

Comment:
Because `join_f` receives both `ValueAndMeta` with supposedly static lifetimes, there are a myriad ways this could be abused and lead to UB. Simplest example, a user calls clone and keeps the ValueAndMeta around -> user after free. I'm sure how to express this ownership transfer into another `ValueAndMeta` while assuming that the thing you originally borrow from is still around, short of actually storing them in Vec. So I suggest at least marking this function `unsafe` and explaining the invariants in great detail.

Reply:
That's a good observation, it requires a bit of a contrived example to cause it to violate the guarantees but this was possible:
```rust
    let vec1 = br#"{"key": "value"}"#.to_vec();
    let mut e1 = EventPayload::new(vec1, |d| tremor_value::parse_to_value(d).unwrap().into());
    let vec2 = br#"{"snot": "badger"}"#.to_vec();
    let e2 = EventPayload::new(vec2, |d| tremor_value::parse_to_value(d).unwrap().into());
    let mut v = Value::null();
    // We try to move the data ov v2 outside of this closure to trick the borrow checker
    // into letting us have it even if it's referenced data no longer exist
    e1.consume::<Error,_>(e2, |v1, v2| {
      let (v2,_) = v2.into_parts();
      v = v2;
      Ok(())
    }).unwrap();
    drop(e1);
    println!("v: {}", v);
```

I changed the signature to:

```rust
    pub fn consume<E, F>(
        &mut self,
        mut other: EventPayload,
        join_f: F,
    ) -> std::result::Result<(), E>
    where
        E: std::error::Error,
        F: for<'iref, 'head> FnOnce(
            &'iref mut ValueAndMeta<'head>,
            ValueAndMeta<'head>,
        ) -> std::result::Result<(), E>,
```

which solves the issue and fails on the example given